### PR TITLE
Add deep copy to IM

### PR DESCRIPTION
**Pull request by PooyaHekmati**
The entire codebase uses data stored in the IM pool. Since Python passes data by reference only, this means that the whole code base has write access to the IM pool. Consequently, if somewhere in the code we retrieve data from IM and change it, other parts of the code will be affected, which is not desirable.
To mitigate this issue, this PR adds deep copy to the get_data() and get_metadata() functions. Deep copy copies all elements and values within a nested data structure. It adds some computation overhead, but it is needed to prevent side effects. 
For ref, deep copy is implemented here: https://github.com/python/cpython/blob/3.11/Lib/copy.py#L128

### CODE REVIEW

File: RUFAS/input_manager.py
Comment (Isgotkowitz) [inline RUFAS/input_manager.py:1]: LGTM, only question is testing-related: should the addresses of the values returned be checked to make sure they are not the same as the data contained in the `InputManager`?

If not necessary, I'll approve.

**Review by allisterakun — APPROVED** [review]
Logic is straightforward, LGTM. Just the same concern about testing that Ed menetioned.

File: RUFAS/input_manager.py
Comment (PooyaHekmati) [reply RUFAS/input_manager.py:1]: I'm not sure if I'm understanding your question/comment correctly. Can you please clarify what you mean?

**Review by JoeWaddell — APPROVED** [review]

File: RUFAS/input_manager.py
Comment (PooyaHekmati) [reply RUFAS/input_manager.py:1]: @Isgotkowitz I'm going to merge this PR as it doesn't change anything about the validation. We can revisit your comment when you are available.

**Pull request merged**